### PR TITLE
kernel/cpu: overhaul per-CPU borrow checking

### PR DESCRIPTION
**Pull request by msft-jlange**
This PR changes the implementation of CPU borrow checking in two important ways.  First, it uses `RefCell` instead of a custom implementation of borrow checking - this is valuable because it relies on core common code.  Second, borrow checking is now performed in all builds, not just debug builds - this is important because the memory safety guarantees made by Rust are intended to apply in all cases, in all code paths, not just those code paths that were validated using debug builds.

The contents of `PerCpu` have been refactored to enforce meaningful borrow checking according to a specific set of principles.
- In general, *all* references to a structure must enforce borrow checking rules to prevent conflicts between shared and mutable references.
- When references are synthesized from global constants (and thus the compiler cannot correctly check borrowed references), borrow checking must be performed dynamically.
- Borrow checking can safely be avoided if, and only if, references to a structure can only be issued as shared references, because no borrow conflict can arise when no mutable references are possible.
- If an object cannot adhere to these borrow checking rules, because of some runtime condition that could require access to object fields in violation of the borrow rules, references to the object must be made using raw pointers.  This requires the use of unsafe code by any logic that has to reach in, and should thus prompt careful justification that any non-compliant reference is actually correct and appropriate.

### CODE REVIEW

**Comment by joergroedel** [comment]
There is Linux kernel boot failure with these changes, I am currently invesigating:

```
[    0.997422][    T1] smpboot: Max logical packages: 1
[    0.998545][    T1] smpboot: Total of 24 processors activated (127199.90 BogoMIPS)
[SVSM] ERROR: Panic: CPU[19] panicked at kernel/src/cpu/percpu.rs:181:9:
assertion failed: self.vmsa.is_some()
[SVSM] ---BACKTRACE---:
[SVSM]   [0xffffff8000132043]
[SVSM]   [0xffffff800016fc3f]
[SVSM]   Invalid frame
[SVSM] ---END---
```

**Comment by joergroedel** [comment]
It looks like there is a pointer offset of 32 bytes between when the VMSAs are set and when they are fetched. The printed addresses are of GuestVmsaRef structures:

```
[SVSM] Setting VMSA and CAA for ptr 0xffffff8000b97000
[SVSM] Set new VMSA and CAA for CPU 23
...
[SVSM] Getting VMSA ptr 0xffffff8000b97020
[SVSM] ERROR: Panic: CPU[23] panicked at kernel/src/cpu/percpu.rs:185:9:
assertion failed: self.vmsa.is_some()
[SVSM] ---BACKTRACE---:
[SVSM]   [0xffffff800013285a]
[SVSM]   [0xffffff800017066f]
[SVSM]   Invalid frame
[SVSM] ---END---
```

File: kernel/src/cpu/percpu.rs [hunk 1]
```diff
@@ -255,75 +324,51 @@ pub struct PerCpu {
...
 
-            let shared_vaddr = vaddr + private_size;
-            let percpu_shared = shared_vaddr.as_mut_ptr::<PerCpuShared>();
-            (*percpu_shared) = PerCpuShared::new();
-
-            let percpu = vaddr.as_mut_ptr::<PerCpu>();
-
-            (*percpu) = PerCpu::new(apic_id, &*percpu_shared);
+            let percpu_unsafe = vaddr.as_mut_ptr::<PerCpuUnsafe>();
+            (*percpu_unsafe) = PerCpuUnsafe::new(apic_id, percpu_unsafe);
```
Comment (joergroedel) [inline kernel/src/cpu/percpu.rs:358]: This creates a confusing imbalance. `alloc()` is an associated method of `struct PerCpu`, but does allocate a `struct PerCpuUnsafe` and returns an inner member of it.
So while at it, the allocation of `PerCpuUnsafe` data structures should be moved into a generic function.

Comment (joergroedel) [inline kernel/src/cpu/percpu.rs:358]: This creates a confusing imbalance. `alloc()` is an associated method of `struct PerCpu`, but does allocate a struct `PerCpuUnsafe` and returns an inner member of it.
So while at it, the allocation of `PerCpuUnsafe` data structures should be moved into a generic function.

Comment (joergroedel) [reply kernel/src/cpu/percpu.rs:358]: This change also needs a corresponding change in `PerCpuAreas::get()`, like mentioned in my other comment on the PR.

**Review by joergroedel — CHANGES_REQUESTED** [review]

**Comment by msft-jlange** [comment]
> Found the problem: The percpu area now holds a PerCpuUnsafe instance, but the `PerCpuAreas::get()` method casts the pointer into a `PerCpuShared` instance. Rewriting `get()` like below fixes the issue:
> [...]
> In general the storing of a `VirtAddr` in `PerCpuInfo` was a bad idea to begin with, it made this mistake slip through undetected. The percpu pointer should be stored in a type-safe way going forward.

Thanks for taking the time to debug this!  I've changed the `PerCpuInfo` to store a reference instead of a pointer or address, so we should now have the type safety we want.

File: kernel/src/cpu/percpu.rs (hunk 1 above)
Comment (msft-jlange) [reply kernel/src/cpu/percpu.rs]: I've restructured the `alloc` methods of `PerCpu` and `PerCpuUnsafe` which should eliminate the confusion.  Callers will still call `PerCpu::alloc()` but there is a more rational chain of allocation and construction to build up the various components.

**Review by joergroedel — APPROVED** [review]
Looks good now. Testing also went well.

**Pull request merged**